### PR TITLE
Use search engine relevance settings in _search

### PR DESCRIPTION
**Pull request by afoucret**
- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against main? Unless there is a good reason otherwise, we prefer pull requests against main and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

### CODE REVIEW

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java [hunk 1]
```diff
@@ -35,96 +37,182 @@
...
-    public Query rewriteQuery(RelevanceMatchQueryBuilder relevanceMatchQueryBuilder, SearchExecutionContext context) throws IOException {
-        Map<String, Float> fieldsAndBoosts = retrieveFieldsAndBoosts(relevanceMatchQueryBuilder.getRelevanceSettingsId(), context);
-        final QueryBuilder combinedFieldsBuilder = new CombinedFieldsQueryBuilder(relevanceMatchQueryBuilder.getQuery(), fieldsAndBoosts);
+    public QueryBuilder rewriteQuery(RelevanceMatchQueryBuilder relevanceMatchQuery, SearchExecutionContext context) throws IOException {
+        final QueryBuilder organicQuery = organicQueryRewriter.rewriteQuery(null, relevanceMatchQuery, context);
+        return curationQueryRewriter.rewriteQuery(organicQuery, relevanceMatchQuery, context);
+    }
 
-        return applyCurations(combinedFieldsBuilder, relevanceMatchQueryBuilder).toQuery(context);
+    private interface QueryRewriter {
```
Comment (kderusso) [inline x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java:58]: As a matter of preference, do we want the interface/abstract class in their own files?

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java [hunk 2]
```diff
@@ -35,96 +37,182 @@
...
+        }
 
-            BoolQueryBuilder excludedDocsBuilder = new BoolQueryBuilder();
-            for (CurationSettings.DocumentReference excludedDoc : curationSettings.excludedDocs()) {
-                BoolQueryBuilder mustQueryBuilder = new BoolQueryBuilder();
-                mustQueryBuilder.must(new TermsQueryBuilder("_index", excludedDoc.index()));
-                mustQueryBuilder.must(new TermsQueryBuilder("_id", excludedDoc.id()));
+        protected S getSettings(RelevanceMatchQueryBuilder relevanceMatchQuery, SearchExecutionContext context) {
+            String settingsId = getSettingsId(relevanceMatchQuery);
+            SearchEngine searchEngine = getSearchEngine(context);
```
Comment (kderusso) [inline x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java:73]: Would it make sense to quickly check for the existence of a searchEngine first and return if it exists, _then_ check for settings/get settings? That may make it clearer that the existence of a searchEngine is a prerequisite check.

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java [hunk 3]
```diff
@@ -35,96 +37,182 @@
...
-                .stream()
-                .map(docRef -> new PinnedQueryBuilder.Item(docRef.index(), docRef.id()))
-                .toArray(PinnedQueryBuilder.Item[]::new);
-            queryBuilder = new PinnedQueryBuilder(queryBuilder, items);
+        protected abstract String getSettingsId(SearchEngine searchEngine);
+
+        protected abstract String getSettingsId(RelevanceMatchQueryBuilder relevanceMatchQuery);
+
+        private SearchEngine getSearchEngine(SearchExecutionContext context) {
+            if (context.getSearchEngineName() == null) {
```
Comment (kderusso) [inline x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java:95]: Consider creating a local variable `searchEngineName` here.

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java [hunk 4]
```diff
@@ -35,96 +37,182 @@
...
-    private Map<String, Float> retrieveFieldsAndBoosts(String relevanceSettingsId, SearchExecutionContext context) {
-        Map<String, Float> fieldsAndBoosts;
-        if (relevanceSettingsId != null) {
-            try {
-                RelevanceSettings relevanceSettings = relevanceSettingsService.getSettings(relevanceSettingsId);
-                QueryConfiguration queryConfiguration = relevanceSettings.getQueryConfiguration();
-                fieldsAndBoosts = queryConfiguration.getFieldsAndBoosts();
-            } catch (AbstractSettingsService.SettingsServiceException e) {
-                throw new IllegalArgumentException("[relevance_match] " + e.getMessage());
+    private static class OrganicQueryRewriter extends AbstractQueryRewriter<RelevanceSettings> {
```
Comment (kderusso) [inline x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java:105]: Why did we choose the name `OrganicQueryRewriter`? Is this performing the function of a `RelevanceQueryRewriter`?

**Review by kderusso — APPROVED** [review]
Overall changes LGTM

File: x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java [hunk 5]
```diff
@@ -35,96 +37,182 @@
...
-    private QueryBuilder applyCurations(QueryBuilder queryBuilder, RelevanceMatchQueryBuilder relevanceMatchQueryBuilder) {
-        String curationsSettingsId = relevanceMatchQueryBuilder.getCurationsSettingsId();
-        if (curationsSettingsId != null) {
-            try {
-                CurationSettings curationSettings = curationsService.getSettings(curationsSettingsId);
-                final boolean conditionMatch = curationSettings.conditions().stream().anyMatch(c -> c.match(relevanceMatchQueryBuilder));
-                if (conditionMatch) {
-                    queryBuilder = applyPinnedDocs(queryBuilder, curationSettings);
-                    queryBuilder = applyExcludedDocs(queryBuilder, curationSettings);
+    private static class CurationQueryRewriter extends AbstractQueryRewriter<CurationSettings> {
```
Comment (carlosdelest) [inline x-pack/plugin/relevance-search/src/main/java/org/elasticsearch/xpack/relevancesearch/query/RelevanceMatchQueryRewriter.java:156]: I think it makes sense to separate these classes to their own class file; I think we will be expanding them in a few iterations 😉

**Review by carlosdelest — APPROVED** [review]
The refactoring LGTM. Some comments for your consideration about structure 👍

**Pull request merged**